src/DocumentDB.ChangeFeedProcessor.IntegrationTests/V3CompatibilityTests.cs

```diff
@@ -0,0 +1,226 @@
+                    .WithLeaseCollection(this.LeaseCollectionInfo)
+                    .BuildAsync();
+
+            await changeFeedProcessorBuilder.StartAsync();
+            await Task.Delay(5000);
+            await changeFeedProcessorBuilder.StopAsync();
+
+            // Verify that no leases have LeaseToken (V3 contract)
+            using (DocumentClient client = new DocumentClient(this.LeaseCollectionInfo.Uri, this.LeaseCollectionInfo.MasterKey, this.LeaseCollectionInfo.ConnectionPolicy))
+            {
+                Uri collectionUri = UriFactory.CreateDocumentCollectionUri(this.LeaseCollectionInfo.DatabaseName, this.LeaseCollectionInfo.CollectionName);
+
+                IDocumentQuery<JObject> query = client.CreateDocumentQuery<JObject>(collectionUri, "SELECT * FROM c").AsDocumentQuery();
+                while (query.HasMoreResults)
+                {
+                    foreach (JObject lease in await query.ExecuteNextAsync())
+                    {
+                        string leaseId = lease.Value<string>("id");
+                        if (leaseId.Contains(".info") || leaseId.Contains(".lock"))
+                        {
+                            // These are the store initialization marks
+                            continue;
+                        }
+
+                        Assert.NotNull(lease.Value<string>("PartitionId"));
+                        Assert.Null(lease.Value<string>("LeaseToken"));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// When the user migrates from V3 CFP, the leases contain LeaseToken.
+        /// To allow for backward compatibility (V2 -> V3 -> V2) we need to honor the existence of LeaseToken and maintain its value in case the lease was created by V3.
+        /// </summary>
+        [Fact]
+        public async Task Schema_OnV2MigrationMaintainLeaseToken()
+        {
+            List<int> expectedIds = Enumerable.Range(0, 20).ToList();
+            List<int> receivedIds = new List<int>();
+            TestObserverFactory observerFactory = new TestObserverFactory(
+                (FeedProcessing.IChangeFeedObserverContext context, IReadOnlyList<Document> docs) =>
+                {
+                    foreach (Document doc in docs)
+                    {
+                        receivedIds.Add(int.Parse(doc.Id));
+                    }
+
+                    return Task.CompletedTask;
+                });
+
+            IChangeFeedProcessor changeFeedProcessorBuilder = await new ChangeFeedProcessorBuilder()
+                    .WithObserverFactory(observerFactory)
+                    .WithHostName("smoke_test")
+                    .WithFeedCollection(this.MonitoredCollectionInfo)
+                    .WithLeaseCollection(this.LeaseCollectionInfo)
+                    .BuildAsync();
+
+            await changeFeedProcessorBuilder.StartAsync();
+            await Task.Delay(10000);
```

Comment:
There is a number of these waits. Can we use more deterministic way? Are we waiting for just lease store init completed or more?

Reply:
On start we could potentially wait for number of unique (ignoring multipple notifications on same partition if any, just in case) OpenAsync notifications in TestObsereverFactory to be equal to partition count (which is the number of lease docs after .info appeared and .lock disappeared. For close we could wait for unique CloseAsync count

Reply:
Removed, used ManualResetEvent